File: files/ScoreComparer.vb
Imports System.ServiceModel.Syndication

Public Class ScoreComparer
    Implements System.Collections.Generic.IComparer(Of SyndicationItem)
    Public Function Compare(ByVal x As SyndicationItem, ByVal y As SyndicationItem) As Integer Implements IComparer(Of System.ServiceModel.Syndication.SyndicationItem).Compare


        Dim higherScoreOne As Integer = _
            GetHigherScore(DirectCast(XElement.ReadFrom(x.ElementExtensions(0).GetReader()), XElement).Value, _
                            DirectCast(XElement.ReadFrom(x.ElementExtensions(1).GetReader()), XElement).Value) '(RTScore, MetaScore)
        Dim higherScoreTwo As Integer = _
            GetHigherScore(DirectCast(XElement.ReadFrom(y.ElementExtensions(0).GetReader()), XElement).Value, _
                            DirectCast(XElement.ReadFrom(y.ElementExtensions(1).GetReader()), XElement).Value) '(RTScore, MetaScore)

        If higherScoreOne > higherScoreTwo Then
            Return 1
        ElseIf higherScoreOne = higherScoreTwo Then
            Return 0
        Else
            Return -1
        End If

    End Function

    Private Function GetHigherScore(ByVal x As String, ByVal y As String) As Integer
        'initialized to zeros
        Dim higherScore As Integer
        Dim numberOne, numberTwo As Integer

        Int32.TryParse(x, numberOne)
        Int32.TryParse(y, numberTwo)

        'could be zeros
        If numberOne > numberTwo Then
            higherScore = numberOne
        Else
            higherScore = numberTwo
        End If

        Return higherScore
    End Function


End Class

File: Gregatr.Domain/Services/ScoreComparer.cs
using System;
using System.ServiceModel.Syndication;
using System.Xml.Linq;

namespace GreGatr.Domain.Services
{
    public class ScoreComparer : IComparer<SyndicationItem>
    {
        public int Compare(SyndicationItem? itemOne, SyndicationItem? itemTwo)
        {
            //if (x == null || y == null) return 0; 
            //for now we want to throw an exception, if somehow item(s) are null
            int higherScoreOne = GetHigherScore(itemOne!);
            int higherScoreTwo = GetHigherScore(itemTwo!);

            //explicit results instead of return higherScoreOne.CompareTo(higherScoreTwo); 
            //default sort order with ICompare.Compare (and CompareTo in comment above) is acsending (smaller value on top) i.e: if (higherScoreOne < higherScoreTwo) return -1;
            //for descending order flip the logic:
            if (higherScoreOne > higherScoreTwo)
                return -1;
            else if (higherScoreOne == higherScoreTwo)
                return 0;
            else 
                return 1; //(higherScoreOne < higherScoreTwo)
        }   

        private static int GetHigherScore(SyndicationItem item)
        {
            int higherScore = 0;
            foreach (var scoreExtension in item.ElementExtensions)
            {
                string scoreStr =((XElement)(XElement.ReadFrom(scoreExtension.GetReader()))).Value;
                int.TryParse(scoreStr, out int score);

                if (score > higherScore) higherScore = score;
            }
            return higherScore;
        }
    }
}

